### PR TITLE
[Merged by Bors] - ET-3369 implement delete documents with actix (9)

### PR DESCRIPTION
**Pull request by rustonaut**
Implement single and batch document deletion on ingestion endpoint.

**References:**

- [ET-3369]
- based on:
  - #673   
  - #672 
  - #671   
  - #670 
  - #669
  - #668
  - #667
  - #659

[ET-3369]: https://xainag.atlassian.net/browse/ET-3369?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by rustonaut** [comment]
I forgot to uncherry-pick some changes, will be fixed in a moment.

**Comment by rustonaut** [comment]
done, now the changes unrelated to delete documents are moved to the PR they should be in

File: discovery_engine_core/web-api2/src/elastic.rs [hunk 1]
```diff
@@ -65,18 +76,97 @@ impl Default for Config {
...
+                    }
+                } else {
+                    error!("Bulk delete request contains non delete responses: {response:?}");
+                }
+            }
+
+            if !errors.is_empty() {
+                return Err(FailedToDeleteSomeDocuments { errors }.into());
+            }
+        }
```
Comment (rustonaut) [inline discovery_engine_core/web-api2/src/elastic.rs:131]: :gray_exclamation: I'm really not sure if having that error handling makes sense.

- Non existing documents don't cause `response.errors`  to be `true`.
- The only way we can have some but not all document deletions failing are (AFIK) supper edge case failure cases in which it often 1) isn't possible to fix it by directly retrying 2) it isn't too big of a problem to retry all once the problem was fixed.

File: discovery_engine_core/web-api2/src/elastic.rs [hunk 2]
```diff
@@ -123,3 +213,15 @@ impl ElasticSearchClient {
...
+#[derive(Deserialize)]
+struct BulkResponse {
+    errors: bool,
+    items: Vec<HashMap<String, BulkItemResponse>>,
+}
+
+#[derive(Debug, Deserialize)]
+struct BulkItemResponse {
+    status: u16,
+    error: Value,
```
Comment (biernacki) [inline discovery_engine_core/web-api2/src/elastic.rs:226]: is error always in the response? I know the ingestion request we used an Option there
```suggestion
    error: Option<Value>,
```

**Review by biernacki — COMMENTED** [review]
left some minor comments

File: discovery_engine_core/web-api2/src/elastic.rs (hunk 2 above)
Comment (rustonaut) [reply discovery_engine_core/web-api2/src/elastic.rs:226]: yes, through it's only missing a `#[serde(default)]` but we don't need to wrap it into an option as there is no semantic difference between not given and set to `null`.

File: discovery_engine_core/web-api2/src/elastic.rs [hunk 3]
```diff
@@ -65,18 +76,97 @@ impl Default for Config {
...
+            )
+            .await?;
+
+        if response.errors {
+            let mut errors = HashMap::new();
+            for (idx, mut response) in response.items.into_iter().enumerate() {
+                if let Some(response) = response.remove("delete") {
+                    if let Ok(status) = StatusCode::from_u16(response.status) {
+                        if status != StatusCode::NOT_FOUND && !status.is_success() {
+                            errors.insert(documents[idx].clone(), response.error);
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply discovery_engine_core/web-api2/src/elastic.rs]: The order is fixed so both approaches work. I changed it to use the `id` field.

File: discovery_engine_core/web-api2/src/elastic.rs [hunk 4]
```diff
@@ -65,18 +76,101 @@ impl Default for Config {
...
+            for mut response in response.items.into_iter() {
+                if let Some(response) = response.remove("delete") {
+                    if let Ok(status) = StatusCode::from_u16(response.status) {
+                        if status != StatusCode::NOT_FOUND && !status.is_success() {
+                            error!(document_id=%response.id, error=%response.error);
+                            errors.push(response.id);
+                        }
+                    } else {
+                        error!("Non http status code: {}", response.status);
+                    }
```
Comment (biernacki) [inline discovery_engine_core/web-api2/src/elastic.rs:123]: Not sure if we should add `DocumentId` to errors based on status being a valid http status, but rather based on the presence of `error` field in `BulkItemResponse`, wdyt?

**Review by biernacki — APPROVED** [review]

File: discovery_engine_core/web-api2/src/elastic.rs (hunk 4 above)
Comment (biernacki) [reply discovery_engine_core/web-api2/src/elastic.rs:123]: >the documentation does give guarantees

Let's leave it for now then. We can see later if we can trust their documentation 😉 

>I think we should just use the official rust bindings even if they still are alpha

Let's maybe put a pin in it until Andrea is back. This was an ad hoc decision that was made when the api was a PoC stage, so we can definitely discuss switching to official bindings I think.

**Comment by rustonaut** [comment]
bors merge

**Comment by rustonaut** [comment]
spurious failure in dart :roll_eyes:

**Comment by rustonaut** [comment]
bors merge

**Comment by rustonaut** [comment]
gitthub passing to ci-ok that a task failed while also reporting all other tasks as not having failed :face_exhaling:

**Comment by rustonaut** [comment]
bors merge

**Comment by rustonaut** [comment]
bors merge

**Pull request closed without merging**